python/src/iceberg/partition_spec.py

```diff
@@ -0,0 +1,289 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from urllib import parse
+from collections import defaultdict
+from typing import List, Optional
+
+from iceberg.partition_field import PartitionField
+from iceberg.schema import Schema
+from iceberg.types import Type, NestedField
+from iceberg.validation_exception import ValidationException
+
+PARTITION_DATA_ID_START = 1000
+
+
+class PartitionSpec(object):
+    fields_by_source_id: defaultdict[list] = None
+    field_list: List[PartitionField] = None
+
+    def __init__(
+        self,
+        schema: Schema,
+        spec_id: int,
+        part_fields: List[PartitionField],
+        last_assigned_field_id: int,
+    ):
+        self._schema = schema
+        self._spec_id = spec_id
+        self._fields = [part_field for part_field in part_fields]
+        self._last_assigned_field_id = last_assigned_field_id
+
+    @property
+    def schema(self) -> Schema:
+        return self._schema
+
+    @property
+    def spec_id(self) -> int:
+        return self._spec_id
+
+    @property
+    def fields(self) -> List[PartitionField]:
+        return self._fields
+
+    @property
+    def last_assigned_field_id(self) -> int:
+        return self._last_assigned_field_id
+
+    def is_partitioned(self):
+        return len(self.fields) < 1
+
+    def _generate_fields_by_source_id(self):
+        if self.fields_by_source_id is None:
+            fields_source_to_field_dict = defaultdict(list)
+            for field in self.fields:
+                fields_source_to_field_dict[field.source_id] = [field]
+            return fields_source_to_field_dict
+        return None
+
+    def get_fields_by_source_id(self, field_id: int) -> List[PartitionField]:
+        return self._generate_fields_by_source_id().get(field_id, None)
```

Comment:
Hi I'm new to iceberg but this signature looks confusing. Is `source_id` and `field_id` the same thing?

```suggestion
    def get_fields_by_source_id(self, field_id: int) -> Optional[List[PartitionField]]:
        return self._generate_fields_by_source_id().get(field_id, None)
```
The returns should be `Optional[List[PartitionField]]` instead if the the method will returns `None`.

I think this method may not be correct because `_generate_fields_by_source_id` will only returns `fields_source_to_field_dict` the first it's called, and all `None` afterward because the `if self.fields_by_source_id is None:` check.

Reply:
Alternatively, you can use `@cached_property` to simplify the `if-else` check and be more pythonic.

```python
@cached_property
def fields_by_source_id(self):
    fields_source_to_field_dict = defaultdict(list)
        for field in self.fields:
            fields_source_to_field_dict[field.source_id] = [field]
        return fields_source_to_field_dict
```